File: src/Libraries/Core/src/Misc/PasswordEncoder.cs
using System;
using System.Collections.Generic;
using System.Text;
using UniSpy.Server.Core.Abstraction.BaseClass;
using UniSpy.Server.Core.Encryption;
using UniSpy.Server.Core.Extension;

namespace UniSpy.Server.Core.Misc
{
    public class PasswordEncoder
    {
        /// <summary>
        /// process password to string which stores in our database
        /// </summary>
        /// <param name="dict"></param>
        public static string ProcessPassword(Dictionary<string, string> dict)
        {
            string md5Password;
            if (dict.ContainsKey("passwordenc"))
            {
                //we decoded gamespy encoded password then get md5 of it         
                md5Password = StringExtensions.GetMD5Hash(Decode(dict["passwordenc"]));
            }
            else if (dict.ContainsKey("passenc"))
            {
                //we decoded gamespy encoded password then get md5 of it  
                md5Password = StringExtensions.GetMD5Hash(Decode(dict["passenc"]));
            }
            else if (dict.ContainsKey("pass"))
            {
                md5Password = StringExtensions.GetMD5Hash(dict["pass"]);
            }
            else if (dict.ContainsKey("password"))
            {
                md5Password = StringExtensions.GetMD5Hash(dict["password"]);
            }
            else
            {
                throw new UniSpyException("Can not find password field in request");
            }
            return md5Password;
        }
        /// <summary>
        /// Encodes a password to Gamespy format
        /// </summary>
        /// <param name="Password"></param>
        /// <returns></returns>
        private static string Encode(string Password)
        {
            // Get password string as UTF8 String, Convert to Base64
            byte[] PasswordBytes = Encoding.UTF8.GetBytes(Password);
            string Pass = Convert.ToBase64String(GameSpyEncodeMethod(PasswordBytes));

            // Convert Standard Base64 to Gamespy Base 64
            StringBuilder builder = new StringBuilder(Pass);
            builder.Replace('=', '_');
            builder.Replace('+', '[');
            builder.Replace('/', ']');
            return builder.ToString();
        }

        /// <summary>
        /// Decodes a Gamespy encoded password
        /// </summary>
        /// <param name="Password"></param>
        /// <returns></returns>
        private static string Decode(string password)
        {
            // Convert Gamespy Base64 to Standard Base 64
            password = password.Replace('_', '=').Replace('[', '+').Replace(']', '/');
            // Decode passsword
            byte[] passwordBytes = Convert.FromBase64String(password);
            return UniSpyEncoding.GetString(GameSpyEncodeMethod(passwordBytes));
        }

        /// <summary>
        /// Gamespy's XOR method to encrypt and decrypt a password
        /// </summary>
        /// <param name="pass"></param>
        /// <returns></returns>
        private static byte[] GameSpyEncodeMethod(byte[] pass)
        {
            int a = 0;
            int num = 0x79707367; // gamespy
            for (int i = 0; i < pass.Length; ++i)
            {
                num = GameSpyByteShift(num);
                a = num % 0xFF;
                pass[i] ^= (byte)a;
            }

            return pass;
        }

        /// <summary>
        /// Not exactly sure what this does, but i know its used to 
        /// reverse the encryption and decryption of a string
        /// </summary>
        /// <param name="num"></param>
        /// <returns></returns>
        private static int GameSpyByteShift(int num)
        {
            int c = (num >> 16) & 0xffff;
            int a = num & 0xffff;

            c *= 0x41a7;
            a *= 0x41a7;
            a += ((c & 0x7fff) << 16);

            if (a < 0)
            {
                a &= 0x7fffffff;
                a++;
            }

            a += (c >> 15);

            if (a < 0)
            {
                a &= 0x7fffffff;
                a++;
            }

            return a;
        }
    }
}

File: src/library/src/extentions/password_encoder.py
import hashlib
import base64

from library.src.exceptions.error import UniSpyException


def process_password(request: dict):
    """process password in standard format and return the password"""
    assert isinstance(request, dict)
    if "passwordenc" in request:
        md5_password = get_md5_hash(decode(request["passwordenc"]))
    elif "passenc" in request:
        md5_password = get_md5_hash(decode(request["passenc"]))
    elif "pass" in request:
        md5_password = get_md5_hash(request["pass"])
    elif "password" in request:
        md5_password = get_md5_hash(request["password"])
    else:
        raise UniSpyException("Can not find password field in request")
    return md5_password


def encode(password: str):
    assert isinstance(password, str)
    password_bytes = password.encode("utf-8")
    pass_encoded = base64.b64encode(game_spy_encode_method(password_bytes))
    pass_encoded = pass_encoded.decode("utf-8")
    pass_encoded = pass_encoded.replace("=", "_").replace("+", "[").replace("/", "]")
    return pass_encoded


def decode(password: str):
    assert isinstance(password, str)
    password = password.replace("_", "=").replace("[", "+").replace("]", "/")
    password_bytes = base64.b64decode(password)
    return game_spy_encode_method(password_bytes).decode("utf-8")


def game_spy_encode_method(password_bytes: bytes):
    assert isinstance(password_bytes, bytes)
    a = 0
    num = 0x79707367  # gamespy
    temp_data = list(password_bytes)
    for i in range(len(password_bytes)):
        num = game_spy_byte_shift(num)
        a = num % 0xFF
        temp_data[i] ^= a
    return bytes(temp_data)


def game_spy_byte_shift(num):
    assert isinstance(num, int)
    c = (num >> 16) & 0xFFFF
    a = num & 0xFFFF

    c *= 0x41A7
    a *= 0x41A7
    a += (c & 0x7FFF) << 16

    if a < 0:
        a &= 0x7FFFFFFF
        a += 1

    a += c >> 15

    if a < 0:
        a &= 0x7FFFFFFF
        a += 1

    return a


def get_md5_hash(data):
    isinstance(data, str)
    md5_hash = hashlib.md5()
    md5_hash.update(data.encode("utf-8"))
    return md5_hash.hexdigest()
